File: racer_template.f
        subroutine load_inputs(n_inputs, lm, sglm, extm, sgextm, target_res)
            implicit none
            double precision lm(%(n_lm)d,0:3,%(n_inputs)d)
            double precision sglm(%(n_sglm)d,0:3,%(n_inputs)d)
            double precision extm(%(n_extm)d,0:3,%(n_inputs)d)
            double precision sgextm(%(n_sgextm)d,0:3,%(n_inputs)d)
            integer n_lm, n_ext, n_sg_ext, n_inputs
            double precision target_res(%(n_inputs)d)
            integer I,J
            double precision buff(0:(%(n_lm)d+%(n_extm)d+%(n_sgextm)d)*4)

            OPEN(967, FILE='./%(input_file_name)s', ERR=976, STATUS='OLD', ACTION='READ')
                READ(967,*,END=978) n_lm, n_ext, n_sg_ext, n_inputs
                DO I=1,n_inputs
                    READ(967,*,END=978) (buff(J),J=0,(n_lm+n_ext+n_sg_ext)*4)
                    DO J=0,n_lm-1
                        lm(J+1,:,I) = buff(J*4:(J+1)*4-1)
                    ENDDO
                    DO J=0,n_lm+n_ext-1
                        sglm(J+1,:,I) = buff(J*4:(J+1)*4-1)
                    ENDDO
                    DO J=n_lm,n_lm+n_ext-1
                        extm(J-n_lm+1,:,I) = buff(J*4:(J+1)*4-1)
                    ENDDO
                    DO J=n_lm+n_ext,n_lm+n_ext+n_sg_ext-1
                        sgextm(J-n_lm-n_ext+1,:,I) = buff(J*4:(J+1)*4-1)
                    ENDDO
                    target_res(I) = buff((n_lm+n_ext+n_sg_ext)*4)
                ENDDO
                GOTO 978
976         CONTINUE
                STOP 'Could not read the inputs.txt data.'
978         CONTINUE
            CLOSE(967)

        end subroutine load_inputs

        subroutine race(lm, sglm, extm, sgextm, res)
            implicit none
            double precision lm(0:%(n_lm)d-1,0:3)
            double precision sglm(0:%(n_sglm)d-1,0:3)
            double precision extm(0:%(n_extm)d-1,0:3)
            double precision sgextm(0:%(n_sgextm)d-1,0:3)
            double precision em(0:%(n_edges)d-1,0:3)
            double precision sd(0:%(n_edges)d-1,0:%(n_edges)d-1)
            double precision em_osE(0:%(n_edges)d-1)
            double precision num, denom
            double precision pi
            parameter (pi=3.141592653589793d0)
            
            double precision res
            integer I,J
            include 'globals.inc'

            if (DBG) then
                do I=0, %(n_lm)d-1
                    write(*,*) "lm(",I,")=",(lm(I,J),J=0,3)
                enddo
                do I=0, %(n_sglm)d-1
                    write(*,*) "sglm(",I,")=",(sglm(I,J),J=0,3)
                enddo
                do I=0, %(n_extm)d-1
                    write(*,*) "extm(",I,")=",(extm(I,J),J=0,3)
                enddo
                do I=0, %(n_sgextm)d-1
                    write(*,*) "sgextm(",I,")=",(sgextm(I,J),J=0,3)
                enddo
            endif

            res = 0.0d0

!           Evaluae all common quantities
%(warmup_code)s
            
!           Now evaluate all cuts
%(evaluate_ltd_cut)s

!           Finalise computation
%(wrapup_code)s

        end subroutine race

        program test_cpu_time
            implicit none
            integer i, j
            integer n_inputs
            double precision lm(%(n_lm)d,0:3,%(n_inputs)d)
            double precision sglm(%(n_sglm)d,0:3,%(n_inputs)d)
            double precision extm(%(n_extm)d,0:3,%(n_inputs)d)
            double precision sgextm(%(n_sgextm)d,0:3,%(n_inputs)d)
            double precision res(%(n_inputs)d), target_res(%(n_inputs)d)
            double precision max_diff
            integer max_diff_index
            real :: start, finish
            integer n_runs
            parameter (n_runs =1)

            include 'globals.inc'

            call load_inputs(n_inputs, lm, sglm, extm, sgextm,target_res)

            call cpu_time(start)
            do i=1,n_runs
                do j=1,n_inputs
                    call race(lm(1,0,j),sglm(1,0,j),extm(1,0,j),sgextm(1,0,j),res(j))
                    if (DBG) then
                        write(*,*) "Result :", res(j)
                        write(*,*) "Target :", target_res(j)
                    endif
                enddo
                if (DBG) then
                    stop
                endif
            enddo
            call cpu_time(finish)
            write(*,*) ((finish-start)*1000000.0d0)/(n_runs*n_inputs)
            max_diff = -1.0d0
            do J=1,n_inputs
                if (dabs((res(j)-target_res(j))/target_res(j)).gt.max_diff) then
                    max_diff = dabs((res(j)-target_res(j))/target_res(j))
                    max_diff_index = I
                endif
            enddo
            write(*,*) max_diff*100.0d0
            write(*,*) 'Max diff of ',max_diff*100.0d0,'%% for input #',max_diff_index,' with res=',res(j),' and target_res=',target_res(j)
        end program test_cpu_time

File: racer_template.py
#!/usr/bin/env python3
import vectors
from math import sqrt, pi
import time
import argparse

_DEBUG = False

def load_inputs(input_path):

    inputs = []
    n_lm, n_ext, n_sg_ext, n_inputs = None, None, None, None
    with open(input_path,'r') as f:
        for i_l,l in enumerate(f.readlines()):
            if i_l == 0:
                n_lm, n_ext, n_sg_ext, n_inputs = [ int(i_str) for i_str in l.strip().split(' ') ]
                continue
            v_is = l.strip().replace('E','e').replace('D','e').replace('d','e').split(' ')
            inputs.append(
                (
                    [vectors.Vector([float(v_i) for v_i in v_is[i_chunk*4:(i_chunk+1)*4]]) for i_chunk in range((len(v_is)-1)//4)],
                    float(v_is[-1])
                )
            )
    return [n_lm, n_ext, n_sg_ext, inputs]

def numerator(em,spatial_dots):
    sd = spatial_dots
%(numerator)s

def evaluate_ltd_cut(i_ltd_cut, lm, extm, sgextm, em, em_osE, spatial_dots):
    sd = spatial_dots
%(evaluate_ltd_cut)s

def set_em_energies(sglm, sgextm, em):

%(set_em_energies)s

def set_em(sglm, sgextm, em):

%(set_em)s


def race(n_lm, n_ext, n_sg_ext, input_data):

    # Naming conventions (all 4D vectors):
    # lm : loop momenta of the amplitude
    # sglm : loop momenta of the supergraph (input from rust)
    # extm : external momenta from the amplitude (but not containing external momenta of the supergraph)
    # sgextm: external momenta from the supergraph (each appears one, not twice for each side of the forward scattering graph)
    # em: all edge momenta of the supergraph
    # spatial_dots: all spatial dots between edge momenta

    # SG loop momenta
    sglm = input_data[:n_lm+n_ext]

    # Amplitude loop momenta
    lm = input_data[:n_lm]
    if _DEBUG:
        print("Amplitude loop momenta:")
        print("-----------------------")
        for i_lm, m in enumerate(lm):
            print('{:<5s} : {:s}'.format('k{:d}'.format(i_lm),','.join('{:>+30.16e}'.format(m_i) for m_i in m)))

    # Amplitude external momenta
    extm = input_data[n_lm:n_lm+n_ext]
    if _DEBUG:
        print("Amplitude external momenta:")
        print("---------------------------")
        for i_m, m in enumerate(extm):
            print('{:<5s} : {:s}'.format('pl{:d}'.format(i_m),','.join('{:>+30.16e}'.format(m_i) for m_i in m)))
    # Supergraph externals
    sgextm = input_data[n_lm+n_ext:]
    if _DEBUG:
        print("Supergraph external momenta:")
        print("-----------------------")
        for i_m, m in enumerate(sgextm):
            print('{:<5s} : {:s}'.format('p{:d}'.format(i_m),','.join('{:>+30.16e}'.format(m_i) for m_i in m)))
    # Amplitude edges momenta
    em = [None,]*%(n_amplitude_edges)s
    set_em(sglm, sgextm, em)
    if _DEBUG:
        mom_id_to_name_map = %(mom_id_to_name_map)s
        print("All edges momenta:")
        print("------------------")
        for i_em, m in enumerate(em):
            print('{:<5s} : {:s}'.format(mom_id_to_name_map[i_em],','.join('{:>+30.16e}'.format(m_i) for m_i in m)))
    em_osE = [None,]*%(n_amplitude_edges)s
    for i_e, m in enumerate(em):
        em_osE[i_e] = sqrt(m[1]*m[1]+m[2]*m[2]+m[3]*m[3])
    spatial_dots = []
    for i, m_i in enumerate(em):
        spatial_dots.append( [ m_i[1]*m_j[1]+m_i[2]*m_j[2]+m_i[3]*m_j[3] for j, m_j in enumerate(em) ] )

    tot = 0.
    for i_ltd_cut in range(%(n_ltd_cuts)d):
        res = evaluate_ltd_cut(i_ltd_cut, lm, extm, sgextm, em, em_osE, spatial_dots)
        if _DEBUG: print("Result from LTD cut #{:d}: {:.16e}".format(i_ltd_cut, res))
        tot += res
    return tot*(%(normalisation)s)

if __name__ == '__main__':

    parser = argparse.ArgumentParser(description="""Python racer.""")

    parser.add_argument('--n_runs', '-n_runs', dest='n_runs', type=int, default=1,
        help='Specify number of runs to average over (default: %%(default)s).')
    parser.add_argument('--debug', '-d', dest='debug', action='store_true', default=False,
        help='Enable debug mode (default: %%(default)s).')
    args = parser.parse_args()

    _DEBUG = args.debug

    n_lm, n_ext, n_sg_ext, inputs  = load_inputs('./%(input_file_name)s')
    
    if _DEBUG:
        n_runs = 1
    else:
        n_runs = args.n_runs

    all_res = []
    all_targets = []
    t_start = time.time()
    for i_r in range(n_runs):
        for i, (input_data, target) in enumerate(inputs):
            res = race(n_lm, n_ext, n_sg_ext, input_data)
            all_res.append(res)
            all_targets.append(target)
            if _DEBUG:
                header="Sample #{:d}:".format(i)
                print("-"*len(header))
                print(header)
                print("-"*len(header))
                print("Result   : {:.16e}".format(res))
                print("Target   : {:.16e}".format(target))
                print("Rel diff : {:.16e}".format(abs((res-target)/res)))
                print("Ratio    : {:.16e}".format(res/target))
    tot_time = time.time()-t_start
    print('{:.16e}'.format(float(tot_time/(n_runs*len(inputs)))*1000000.))
    max_diff = -1.
    max_index = None
    for i, (res, target) in enumerate(zip(all_res,all_targets)):
        if abs((res-target)/target)>max_diff:
            max_diff = abs((res-target)/target)
            max_index = i
    print('{:.5e}'.format(max_diff*100.))
    print("Max difference of {:.3e}%% for input #{:d} with res={:.16e} and target_res={:.16e}".format(
        max_diff*100., max_index, all_res[max_index], all_targets[max_index]
    ))